bson/src/main/org/bson/codecs/pojo/LazyPropertyModelCodec.java

```diff
@@ -0,0 +1,145 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bson.codecs.pojo;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.RepresentationConfigurable;
+import org.bson.codecs.configuration.CodecConfigurationException;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.bson.codecs.pojo.PojoSpecializationHelper.specializeTypeData;
+
+class LazyPropertyModelCodec<T> implements Codec<T> {
+    private final PropertyModel<T> propertyModel;
+    private final CodecRegistry registry;
+    private final PropertyCodecRegistry propertyCodecRegistry;
+    private final DiscriminatorLookup discriminatorLookup;
+
+    private Codec<T> codec;
+
+    LazyPropertyModelCodec(final PropertyModel<T> propertyModel, final CodecRegistry registry,
+                                  final PropertyCodecRegistry propertyCodecRegistry, final DiscriminatorLookup discriminatorLookup) {
+        this.propertyModel = propertyModel;
+        this.registry = registry;
+        this.propertyCodecRegistry = propertyCodecRegistry;
+        this.discriminatorLookup = discriminatorLookup;
+    }
+
+    @Override
+    public T decode(final BsonReader reader, final DecoderContext decoderContext) {
+        return getPropertyModelCodec().decode(reader, decoderContext);
+    }
+
+    @Override
+    public void encode(final BsonWriter writer, final T value, final EncoderContext encoderContext) {
+        getPropertyModelCodec().encode(writer, value, encoderContext);
+    }
+
+    @Override
+    public Class<T> getEncoderClass() {
+        return propertyModel.getTypeData().getType();
+    }
+
+    private synchronized Codec<T> getPropertyModelCodec() {
+        if (codec == null) {
+            Codec<T> localCodec = getCodecFromPropertyRegistry(propertyModel);
+            if (localCodec instanceof PojoCodec) {
+                PojoCodec<T> pojoCodec = (PojoCodec<T>) localCodec;
+                ClassModel<T> specialized = getSpecializedClassModel(pojoCodec.getClassModel(), propertyModel);
+                localCodec = new PojoCodecImpl<>(specialized, registry, propertyCodecRegistry, discriminatorLookup, true);
+            }
+            if (localCodec == null) {
+                localCodec = new LazyMissingCodec<>(propertyModel.getTypeData().getType(),
```

Comment:
I've moved it into the `getCodecFromPropertyRegistry` its not dead code just very defensive. If a user creates their own codec registry implementation that returns null from `registry.get(clazz)` then it would be required.

Reply:
That would violate the specification of the CodecRegistry interface but you never know what users will do.